gcsio/src/main/java/com/google/cloud/hadoop/gcsio/MetricsPublisher.java

```diff
@@ -0,0 +1,243 @@
+      throws IOException {
+    synchronized (monitor) {
+      if (!initialized) {
+        StackdriverStatsExporter.createAndRegister(
+            StackdriverStatsConfiguration.builder()
+                .setCredentials(credentials)
+                .setProjectId(projectId)
+                .setExportInterval(fromMillis(60000))
+                .build());
+        Runtime.getRuntime().addShutdownHook(new Thread(StackdriverStatsExporter::unregister));
+        initialized = true;
+      }
+    }
+  }
+
+  private static void registerAllViews() {
+    // Defining the distribution aggregations in milliseconds
+    Aggregation latencyDistribution =
+        Distribution.create(
+            BucketBoundaries.create(
+                Arrays.asList(
+                    0.0, 25.0, 50.0, 75.0, 100.0, 200.0, 400.0, 600.0, 800.0, 1000.0, 5000.0,
+                    10000.0, 15000.0, 20000.0, 25000.0, 30000.0)));
+
+    // Define the count aggregation
+    Aggregation countAggregation = Aggregation.Count.create();
+
+    // Define the views
+    View[] views =
+        new View[] {
+          View.create(
+              Name.create("gcsio/latency"),
+              "The distribution of latencies for a  method",
+              LATENCY_MS,
+              latencyDistribution,
+              TAG_KEYS),
+          View.create(
+              Name.create("gcsio/message/latency"),
+              "The distribution of latencies at a message level in a streaming context",
+              MESSAGE_LATENCY_MS,
+              latencyDistribution,
+              TAG_KEYS),
+          View.create(
+              Name.create("gcsio/retries"),
+              "The distribution of retry attempts for a method",
+              REQUEST_RETRIES,
+              countAggregation,
+              TAG_KEYS),
+          View.create(
+              Name.create("gcsio/requests"),
+              "The distribution of request counts for a method",
+              REQUESTS,
+              countAggregation,
+              TAG_KEYS),
+        };
+
+    ViewManager viewManager = Stats.getViewManager();
+
+    for (View view : views) {
+      viewManager.registerView(view);
+    }
+
+    // register views for native grpc client metrics
+    RpcViews.registerAllGrpcViews();
+  }
+
+  @Override
+  public void recordTaggedStat(TagKey key, String value, MeasureLong ml, Long n) {
+    TagContext tagContext =
+        tagger
+            .emptyBuilder()
+            .put(
+                KEY_THREAD,
+                TagValue.create(String.valueOf(Thread.currentThread().getId())),
+                TAG_METADATA_NO_PROPAGATION)
+            .put(key, TagValue.create(value), TAG_METADATA_NO_PROPAGATION)
+            .build();
+    try (Scope ignored = tagger.withTagContext(tagContext)) {
+      statsRecorder.newMeasureMap().put(ml, n).record();
+    }
+  }
+
+  @Override
+  public void recordLong(TagKey[] keys, String[] values, MeasureLong ml, Long n) {
```

Comment:
Can we use map instead of keys and values?

Reply:
Unfortunately, the opencensus API accepts arrays of keys and values. Type casting twice would be redundant.